### PR TITLE
Persistent RPC Connections

### PR DESCRIPTION
**Pull request by rgrinberg**
This PR introduces a new `connect_persistent` function, and a corresponding
with flag `--persistent` for `dune rpc init`.

In this mode persistent mode, the client will try to connect to a build server.
After this initial connection terminates (either because the server quit, or
wasn't running) the client will listen for more build sessions. The result is
that editors can just run `dune rpc init --persistent` at any point and need not
worry about reconnecting

## TODO

- [x] Make this work on windows
- [x] Terminate pending requests once a session disconnects
- [x] Disconnect callback

### Other changes:

- I've moved symlink handling/cleanup to csexp_rpc. It makes it easier to get
consistent behavior on this everywhere.

- I've removed the test mode. In the beginning, I had ambitions of writing tests for RPC in cram. To accomplish that a special `--test` mode was added.  However, writing tests in such style proved to be impractical as it required a level of co-ordination that is hard to describe in sh. In the end, RPC isn't a "command line" feature, so we shouldn't force testing to go through cram.

### CODE REVIEW

**Comment by rgrinberg** [comment]
There's still something unsatisfying about the way I'm handling errors in the client library. I'm currently using a mix of custom exceptions and an error type for errors received from the server. I propose that we remove all result types from the public api and settle on raising on 3 exceptions:

* Programming errors - these crash the program and there's no way to recover from them
* Fatal errors - these terminate the client. After such an exception, all pending requests fail and to recover, one must reconnect and create a new session.
* Response errors - These errors do not terminate the session. They are usually caused by invalid input being provided to the server.

File: src/fiber/fiber.mli [hunk 1]
```diff
@@ -298,6 +298,10 @@ module Stream : sig
     val sequential_iter : 'a t -> f:('a -> unit fiber) -> unit fiber
 
     val parallel_iter : 'a t -> f:('a -> unit fiber) -> unit fiber
+
+    val append : 'a t -> 'a t -> 'a t
+
+    val cons : 'a t -> 'a -> 'a t
```
Comment (jeremiedimino) [inline src/fiber/fiber.mli:304]: Having `cons t x` produces a stream that starts with `x` and is followed by elements of `t` is counter-intuitive. On this occasion, I think it's fine to break the rule that `t` comes first. Or the function should be renamed.

**Review by jeremiedimino — CHANGES_REQUESTED** [review]

File: src/fiber/fiber.mli (hunk 1 above)
Comment (rgrinberg) [reply src/fiber/fiber.mli]: I flipped it

**Comment by rgrinberg** [comment]
@cwong-ocaml can you review this one?

**Review by cwong-ocaml — APPROVED** [review]
Looks fine! I've been using this in Jane Street for a while.

File: otherlibs/dune-rpc/dune_rpc.ml [hunk 2]
```diff
@@ -50,5 +50,11 @@ module V1 = struct
       -> Initialize.t
       -> f:(t -> 'a fiber)
       -> 'a fiber
+
+    val connect_persistent :
+         chan
+      -> on_connect:(unit -> ('a * Initialize.t * Handler.t option) fiber)
+      -> on_connected:('a -> t -> unit fiber)
```
Comment (cwong-ocaml) [inline otherlibs/dune-rpc/dune_rpc.ml:57]: These names aren't super descriptive, and it's not immediately obvious what the difference is from the types. Maybe some doc would be good

Comment (rgrinberg) [reply otherlibs/dune-rpc/dune_rpc.ml:57]: I don't like the names either. Do you have some alternative names in mind?

**Comment by rgrinberg** [comment]
@cwong-ocaml there's now a disconnect callback.

**Comment by cwong-ocaml** [comment]
I'm going to merge once CI is finished; the remaining changes aren't super worth bikeshedding over right now.

**Comment by rgrinberg** [comment]
Alright, I think I finally made the return codes on shutdown consistent. Now let's pray for CI.

**Pull request merged**